### PR TITLE
feat: simplify git dependency detection by only using cargo lock

### PR DESCRIPTION
**Pull request by yusdacra**
This PR also adds `allRefs` and `submodules` support. Since these features were added to `builtins.fetchGit` recently, they are not enabled by default.

### CODE REVIEW

**Comment by mschwaig** [comment]
I tested this branch to see if it works for fetching the submodules of https://crates.io/crates/tflite/0.9.5.

It works for those regular crates.io dependencies like this:
```
tflite = "0.9.5"
```
And it also works for git dependencies like this:
```
tflite = { git = "https://github.com/mschwaig/tflite-rs.git", branch = "stop-rewriting-flattbuffers-test" }
```

I think merging this PR would close https://github.com/nmattia/naersk/issues/110 and that well-written issue was really helpful to me.

Before I found this PR I wanted to open a PR for this tiny change to fetch submodules for git dependencies only: https://github.com/mschwaig/naersk/commit/2dd2f1b86d8fe9389430e132045f1ca56f76a184

`tflite` still fails to build with naersk for unrelated reasons, but I tested with that because that's what I'm working on

**Comment by yusdacra** [comment]
> I tested this branch to see if it works for fetching the submodules of [crates.io/crates/tflite/0.9.5](https://crates.io/crates/tflite/0.9.5).
> 
> I set `gitSubmodules = true` but did not set `allRefs = true;`.
> 
> That works for those regular crates.io dependencies like this:
> 
> ```
> tflite = "0.9.5"
> ```
> 
> And the same attributes also work for git dependencies like this:
> 
> ```
> tflite = { git = "https://github.com/mschwaig/tflite-rs.git", branch = "stop-rewriting-flattbuffers-test" }
> ```
> 
> I think merging this PR would close #110 and that well-written issue was really helpful to me.
> 
> Before I found this PR I wanted to open a PR for this tiny change to fetch submodules for git dependencies only: [mschwaig@2dd2f1b](https://github.com/mschwaig/naersk/commit/2dd2f1b86d8fe9389430e132045f1ca56f76a184)
> 
> `tflite` still fails to build with naersk for unrelated reasons, but I tested with that because that's what I'm working on

Unrelated to this PR; but may I ask why and where `tflite-rs` errors?

**Comment by mschwaig** [comment]
> Unrelated to this PR; but may I ask why and where `tflite-rs` errors?

It builds fine with `nix develop`, but fails to link when built with `nix build`.

You can take a closer look here if you are interested: https://github.com/mschwaig/tflite-rs-not-building

That's the minimal test case I got so far. Maybe I should open an issue for this, but I'm not sure. I have not tried building it without `naersk` yet.

**Comment by nrdxp** [comment]
Is there something holding this up? I've been using it fine for over a month now as I need it for direct git dependencies in a work project.

File: README.md
Comment (blitz) [inline README.md:51]: If this maps to `builtins.fetchGit`'s submodule support, we should note here that this only works for Nix 2.4+.

File: README.md
Comment (blitz) [inline README.md:50]: Can you add a test for this functionality? See `tests.nix` for existing tests.

File: lib.nix
Comment (blitz) [inline lib.nix:78]: Is there any functionality that is being lost by not looking into Cargo.toml?

**Review by blitz — CHANGES_REQUESTED** [review]

**Comment by FlorianFranzen** [comment]
I am a big fan of this PR, as it allowed me to only link actual dependencies when creating `git-deps`, which drastically reduces build times for my use case. See  yusdacra/naersk#1 for details.

**Comment by jwoudenberg** [comment]
Just ran into #152. Switching to the `naersk` from this branch fixed things for me.

**Comment by Patryk27** [comment]
Thanks for this patch, it's working great!

I've found only one small issue - something on the way doesn't handle branches with `/` correctly, so e.g. trying to compile a patch such as this:
```toml
[patch."ssh://...."]
foo = { git = "ssh://...", branch = "foo-123/bar" }
```
... will fail on a `cp` on the way.

**Comment by Patryk27** [comment]
@yusdacra thanks! Would you also like to address the comments from @blitz? If you don't have time or something, it's fine - I can address them in a separate merge request myself 🙂

**Comment by Patryk27** [comment]
@yusdacra could you tick `Allow edits from maintainers`? I'll then just push the changes into this pull request and get it merged 🙂

**Comment by yusdacra** [comment]
> @yusdacra could you tick `Allow edits from maintainers`? I'll then just push the changes into this pull request and get it merged 🙂

It seems to already be ticked? Also sorry for not replying, I haven't really had time.

File: README.md
Comment (Patryk27) [reply README.md:50]: I've tried to add a test where `Cargo.lock` refers to a repository with non-master-named branch (which, from what I gathered, is one of the use cases for `allRefs`), but everything seems to be working properly even without the switch; do you maybe have an idea on how to create a test for that?

**Comment by Patryk27** [comment]
I'm having some github-permission-issues trying to push my changes in here, so I'm going to merge this pull request into a temporary, non-master branch, push my changes there, and then create a separate PR for that updated branch, hope you don't mind.

Thank y'all for reviews and see you in the updated PR (I'll post a link here when it's done) 🙂

File: lib.nix
Comment (Patryk27) [reply lib.nix:78]: As far as I've checked, we don't lose anything; I've been using this patched version myself for a couple of past months and I haven't found any issues either.

**Comment by Patryk27** [comment]
Follow-up: https://github.com/nix-community/naersk/pull/231

**Pull request merged**
